File: src/prototype/web/src/input/layerD.ts
// PURE — Layer D: resolves "technique confirm" inputs while the judgment
// window is OPEN. Reference: docs/design/input_system_v1.md §D.1.1 last column.
//
// This module owns the physical-input → Technique mapping that fires only
// while a window is OPEN. A few technique commits are patterns (e.g.
// "R_TRIGGER 0→1 within 200ms", "BTN_BASE hold 500ms") so Layer D keeps a
// small temporal state object. Like Layer B it is a pure transform that
// threads its state through the caller.

import { ButtonBit, type InputFrame } from "./types.js";
import type { HipIntent } from "./intent.js";
import type { Technique } from "../state/judgment_window.js";

// §D.1.1 thresholds — kept central for playtest tuning.
export const LAYER_D_TIMING = Object.freeze({
  triangleHoldMs: 500,        // BTN_BASE long press for triangle
  hipBumpPressWindowMs: 200,  // R_TRIGGER 0→1 must happen inside this window
});

export const LAYER_D_THRESHOLDS = Object.freeze({
  lsHorizontalAbs: 0.8,       // "LS pure horizontal"
  lsUp: 0.8,                  // "LS up"
  rsUp: 0.8,                  // "RS up"
  hipBumpReleasedMaxTrigger: 0.1,
  hipBumpPressedMinTrigger: 0.9,
  crossCollarBothMin: 0.95,   // §D.1.1 "L+R TRIGGER simultaneous max"
});

export type LayerDState = Readonly<{
  // Triangle long-press: frame count BTN_BASE has been held.
  btnBaseHeldMs: number;
  // Hip-bump rapid press tracking: timestamp when R_TRIGGER was most
  // recently seen at ≤ releasedMaxTrigger. If R_TRIGGER then crosses
  // pressedMinTrigger within `hipBumpPressWindowMs`, the commit fires.
  rTriggerLastReleasedMs: number;
  rTriggerPrevValue: number;
  // Cross-collar requires both triggers simultaneously at max — one tick
  // of "both ≥ threshold" is enough; no sustain needed.
}>;

export const INITIAL_LAYER_D_STATE: LayerDState = Object.freeze({
  btnBaseHeldMs: 0,
  rTriggerLastReleasedMs: Number.NEGATIVE_INFINITY,
  rTriggerPrevValue: 0,
});

export type LayerDInputs = Readonly<{
  nowMs: number;
  dtMs: number;
  frame: InputFrame;
  hip: HipIntent;
  // Candidate set currently lit by the judgment window. Layer D only
  // resolves commits for techniques actually in this list — otherwise a
  // pattern match could "confirm" a technique that wasn't a candidate.
  candidates: ReadonlyArray<Technique>;
  // Only active during window state OPEN per §8.1 — caller is responsible
  // for passing false when the window is CLOSED / OPENING / CLOSING.
  windowIsOpen: boolean;
}>;

export function resolveLayerD(
  prev: LayerDState,
  inp: LayerDInputs,
): { next: LayerDState; confirmedTechnique: Technique | null } {
  // Always update the per-frame tracking state, even when the window is
  // closed — otherwise the rapid-press window would start fresh the moment
  // OPENING fires and deny the intended "pre-loaded" hip bump input.
  const nextState = updateTrackingState(prev, inp);

  if (!inp.windowIsOpen) {
    return { next: nextState, confirmedTechnique: null };
  }

  // Resolve in the §D.1.1 table order. Returning on the first match
  // matches the design intent: one technique commits per frame.
  const candSet = new Set(inp.candidates);

  // SCISSOR_SWEEP: L_BUMPER edge + LS horizontal.
  if (
    candSet.has("SCISSOR_SWEEP") &&
    (inp.frame.button_edges & ButtonBit.L_BUMPER) !== 0 &&
    Math.abs(inp.frame.ls.x) >= LAYER_D_THRESHOLDS.lsHorizontalAbs
  ) {
    return { next: nextState, confirmedTechnique: "SCISSOR_SWEEP" };
  }

  // FLOWER_SWEEP: R_BUMPER edge + LS up.
  if (
    candSet.has("FLOWER_SWEEP") &&
    (inp.frame.button_edges & ButtonBit.R_BUMPER) !== 0 &&
    inp.frame.ls.y >= LAYER_D_THRESHOLDS.lsUp
  ) {
    return { next: nextState, confirmedTechnique: "FLOWER_SWEEP" };
  }

  // TRIANGLE: BTN_BASE held ≥ 500ms this frame.
  if (
    candSet.has("TRIANGLE") &&
    nextState.btnBaseHeldMs >= LAYER_D_TIMING.triangleHoldMs
  ) {
    return { next: nextState, confirmedTechnique: "TRIANGLE" };
  }

  // OMOPLATA: L_BUMPER edge + RS up.
  if (
    candSet.has("OMOPLATA") &&
    (inp.frame.button_edges & ButtonBit.L_BUMPER) !== 0 &&
    inp.frame.rs.y >= LAYER_D_THRESHOLDS.rsUp
  ) {
    return { next: nextState, confirmedTechnique: "OMOPLATA" };
  }

  // HIP_BUMP: R_TRIGGER 0→1 within the 200ms window.
  if (
    candSet.has("HIP_BUMP") &&
    inp.frame.r_trigger >= LAYER_D_THRESHOLDS.hipBumpPressedMinTrigger &&
    prev.rTriggerPrevValue < LAYER_D_THRESHOLDS.hipBumpPressedMinTrigger &&
    inp.nowMs - prev.rTriggerLastReleasedMs <= LAYER_D_TIMING.hipBumpPressWindowMs
  ) {
    return { next: nextState, confirmedTechnique: "HIP_BUMP" };
  }

  // CROSS_COLLAR: both triggers simultaneously at max.
  if (
    candSet.has("CROSS_COLLAR") &&
    inp.frame.l_trigger >= LAYER_D_THRESHOLDS.crossCollarBothMin &&
    inp.frame.r_trigger >= LAYER_D_THRESHOLDS.crossCollarBothMin
  ) {
    return { next: nextState, confirmedTechnique: "CROSS_COLLAR" };
  }

  return { next: nextState, confirmedTechnique: null };
}

function updateTrackingState(prev: LayerDState, inp: LayerDInputs): LayerDState {
  const baseHeld = (inp.frame.buttons & ButtonBit.BTN_BASE) !== 0;
  const btnBaseHeldMs = baseHeld ? prev.btnBaseHeldMs + inp.dtMs : 0;

  // Track the most recent "released" timestamp for the hip bump window.
  // If the current value is below the release threshold, remember now.
  const rel = inp.frame.r_trigger <= LAYER_D_THRESHOLDS.hipBumpReleasedMaxTrigger;
  const rTriggerLastReleasedMs = rel ? inp.nowMs : prev.rTriggerLastReleasedMs;

  return Object.freeze({
    btnBaseHeldMs,
    rTriggerLastReleasedMs,
    rTriggerPrevValue: inp.frame.r_trigger,
  });
}

File: src/prototype/unity/Assets/BJJSimulator/Runtime/Input/LayerD.cs
// Ported 1:1 from src/prototype/web/src/input/layerD.ts.
// PURE — Layer D: judgment-window commit resolution.
// Reference: docs/design/input_system_v1.md §D.1.1.
//
// Resolves technique commits only while the judgment window is OPEN.
// Owns a small temporal state for long-press and rapid-press detection.

namespace BJJSimulator
{
    // -------------------------------------------------------------------------
    // Timing / thresholds (§D.1.1)
    // -------------------------------------------------------------------------

    public struct LayerDTiming
    {
        public float TriangleHoldMs;
        public float HipBumpPressWindowMs;

        public static readonly LayerDTiming Default = new LayerDTiming
        {
            TriangleHoldMs       = 500f,
            HipBumpPressWindowMs = 200f,
        };
    }

    public struct LayerDThresholds
    {
        public float LsHorizontalAbs;
        public float LsUp;
        public float RsUp;
        public float HipBumpReleasedMaxTrigger;
        public float HipBumpPressedMinTrigger;
        public float CrossCollarBothMin;

        public static readonly LayerDThresholds Default = new LayerDThresholds
        {
            LsHorizontalAbs          = 0.8f,
            LsUp                     = 0.8f,
            RsUp                     = 0.8f,
            HipBumpReleasedMaxTrigger = 0.1f,
            HipBumpPressedMinTrigger  = 0.9f,
            CrossCollarBothMin       = 0.95f,
        };
    }

    // -------------------------------------------------------------------------
    // Per-player temporal state
    // -------------------------------------------------------------------------

    public struct LayerDState
    {
        public float BtnBaseHeldMs;
        public long  RTriggerLastReleasedMs; // BJJConst.SentinelTimeMs until first release
        public float RTriggerPrevValue;

        public static readonly LayerDState Initial = new LayerDState
        {
            BtnBaseHeldMs           = 0f,
            RTriggerLastReleasedMs  = BJJConst.SentinelTimeMs,
            RTriggerPrevValue       = 0f,
        };
    }

    // -------------------------------------------------------------------------
    // Tick inputs
    // -------------------------------------------------------------------------

    public struct LayerDInputs
    {
        public long        NowMs;
        public float       DtMs;
        public InputFrame  Frame;
        public HipIntent   Hip;
        public Technique[] Candidates;
        public bool        WindowIsOpen;
    }

    // -------------------------------------------------------------------------
    // Pure resolver
    // -------------------------------------------------------------------------

    public static class LayerDOps
    {
        public static (LayerDState NextState, Technique? ConfirmedTechnique) Resolve(
            LayerDState prev,
            LayerDInputs inp,
            LayerDTiming   timing     = default,
            LayerDThresholds thresh   = default)
        {
            if (timing.TriangleHoldMs == 0f)       timing = LayerDTiming.Default;
            if (thresh.LsHorizontalAbs == 0f)      thresh = LayerDThresholds.Default;

            // Always update tracking state (even when window is closed) so
            // rapid-press detection accumulates before OPENING fires.
            var next = UpdateTracking(prev, inp, timing, thresh);

            if (!inp.WindowIsOpen)
                return (next, null);

            var cand = new System.Collections.Generic.HashSet<Technique>(
                inp.Candidates ?? System.Array.Empty<Technique>());

            // Resolve in §D.1.1 table order — first match wins.

            // SCISSOR_SWEEP: LBumper edge + LS pure horizontal
            if (cand.Contains(Technique.ScissorSweep) &&
                (inp.Frame.ButtonEdges & ButtonBit.LBumper) != 0 &&
                System.Math.Abs(inp.Frame.Ls.X) >= thresh.LsHorizontalAbs)
                return (next, Technique.ScissorSweep);

            // FLOWER_SWEEP: RBumper edge + LS up
            if (cand.Contains(Technique.FlowerSweep) &&
                (inp.Frame.ButtonEdges & ButtonBit.RBumper) != 0 &&
                inp.Frame.Ls.Y >= thresh.LsUp)
                return (next, Technique.FlowerSweep);

            // TRIANGLE: BtnBase held ≥ 500ms
            if (cand.Contains(Technique.Triangle) &&
                next.BtnBaseHeldMs >= timing.TriangleHoldMs)
                return (next, Technique.Triangle);

            // OMOPLATA: LBumper edge + RS up
            if (cand.Contains(Technique.Omoplata) &&
                (inp.Frame.ButtonEdges & ButtonBit.LBumper) != 0 &&
                inp.Frame.Rs.Y >= thresh.RsUp)
                return (next, Technique.Omoplata);

            // HIP_BUMP: RTrigger 0→1 within 200ms window
            if (cand.Contains(Technique.HipBump) &&
                inp.Frame.RTrigger >= thresh.HipBumpPressedMinTrigger &&
                prev.RTriggerPrevValue < thresh.HipBumpPressedMinTrigger &&
                prev.RTriggerLastReleasedMs != BJJConst.SentinelTimeMs &&
                inp.NowMs - prev.RTriggerLastReleasedMs <= (long)timing.HipBumpPressWindowMs)
                return (next, Technique.HipBump);

            // CROSS_COLLAR: both triggers simultaneously at max
            if (cand.Contains(Technique.CrossCollar) &&
                inp.Frame.LTrigger >= thresh.CrossCollarBothMin &&
                inp.Frame.RTrigger >= thresh.CrossCollarBothMin)
                return (next, Technique.CrossCollar);

            return (next, null);
        }

        // -----------------------------------------------------------------------

        private static LayerDState UpdateTracking(
            LayerDState prev, LayerDInputs inp,
            LayerDTiming timing, LayerDThresholds thresh)
        {
            bool baseHeld = (inp.Frame.Buttons & ButtonBit.BtnBase) != 0;
            float baseMs  = baseHeld ? prev.BtnBaseHeldMs + inp.DtMs : 0f;

            bool  rel           = inp.Frame.RTrigger <= thresh.HipBumpReleasedMaxTrigger;
            long  lastReleasedMs = rel ? inp.NowMs : prev.RTriggerLastReleasedMs;

            return new LayerDState
            {
                BtnBaseHeldMs          = baseMs,
                RTriggerLastReleasedMs = lastReleasedMs,
                RTriggerPrevValue      = inp.Frame.RTrigger,
            };
        }
    }
}
